x/did/types/did.go

```diff
@@ -0,0 +1,436 @@
+	return func(opts *DIDDocument) {
+		opts.VerificationMethods = verificationMethods
+	}
+}
+
+func WithAuthentications(authentications []*VerificationRelationship) DIDDocumentOption {
+	return func(opts *DIDDocument) {
+		opts.Authentications = authentications
+	}
+}
+
+func WithAssertionMethods(assertionMethods []*VerificationRelationship) DIDDocumentOption {
+	return func(opts *DIDDocument) {
+		opts.AssertionMethods = assertionMethods
+	}
+}
+
+func WithKeyAgreements(keyAgreements []*VerificationRelationship) DIDDocumentOption {
+	return func(opts *DIDDocument) {
+		opts.KeyAgreements = keyAgreements
+	}
+}
+
+func WithCapabilityInvocations(capabilityInvocations []*VerificationRelationship) DIDDocumentOption {
+	return func(opts *DIDDocument) {
+		opts.CapabilityInvocations = capabilityInvocations
+	}
+}
+
+func WithCapabilityDelegations(capabilityDelegations []*VerificationRelationship) DIDDocumentOption {
+	return func(opts *DIDDocument) {
+		opts.CapabilityDelegations = capabilityDelegations
+	}
+}
+
+func WithServices(services []*Service) DIDDocumentOption {
+	return func(opts *DIDDocument) {
+		opts.Services = services
+	}
+}
+
+func (doc DIDDocument) Valid() bool {
+	if doc.Empty() { // deactivated
+		return true
+	}
+
+	if !ValidDID(doc.ID) || doc.VerificationMethods == nil || doc.Authentications == nil {
+		return false
+	}
+
+	if !EmptyDID(doc.Controller) && !ValidDID(doc.Controller) {
+		return false
+	}
+
+	if doc.Contexts == nil || !ValidContexts(doc.Contexts) {
+		return false
+	}
+
+	for _, verificationMethod := range doc.VerificationMethods {
+		if !verificationMethod.Valid(doc.ID) {
+			return false
+		}
+	}
+
+	if !doc.validVerificationRelationships(doc.Authentications) {
+		return false
+	}
+	if !doc.validVerificationRelationships(doc.AssertionMethods) {
+		return false
+	}
+	if !doc.validVerificationRelationships(doc.KeyAgreements) {
+		return false
+	}
+	if !doc.validVerificationRelationships(doc.CapabilityInvocations) {
+		return false
+	}
+	if !doc.validVerificationRelationships(doc.CapabilityDelegations) {
+		return false
+	}
+
+	for _, service := range doc.Services {
+		if !service.Valid() {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (doc DIDDocument) validVerificationRelationships(relationships []*VerificationRelationship) bool {
+	for _, relationship := range relationships {
+		if !relationship.Valid(doc.ID) {
+			return false
+		}
+		if !relationship.hasDedicatedMethod() {
+			// if the relationship isn't a dedicated verification method,
+			// the referenced verification method must be presented in the 'verificationMethod' property.
+			if _, ok := doc.VerificationMethodByID(relationship.VerificationMethodID); !ok {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+
+func (doc DIDDocument) Empty() bool {
+	return EmptyDID(doc.ID)
+}
+
+// TODO Duplicate with protobuf
+/*func (doc DIDDocument) String() string {
+	bz, _ := json.Marshal(doc)
+	return string(bz)
+}*/
```

Comment:
👍 Yeah, right. I think we can remove this comment. This `String()` is not logically important. So, we can use the proto-generated one.